File: backup/Breadth-first_search(BFS)/python/05_미로탈출.py
# 이것이 취업을 위한 코딩테스트다 152p
# https://www.acmicpc.net/problem/2178
# 2

from collections import deque

mx, my = [-1, 1, 0, 0], [0, 0, -1, 1]

def bfs():
    q = deque()
    q.append((0, 0))

    while q:
        px, py = q.popleft()
        if px == N - 1 and py == M - 1:
            return matrix[px][py]

        for i in range(4):
            nx = px + mx[i]
            ny = py + my[i]

            if nx < 0 or nx >= N or ny < 0 or ny >= M:
                continue
            if matrix[nx][ny] == 1:
                matrix[nx][ny] = matrix[px][py] + 1
                q.append((nx, ny))

N, M = map(int, input().split())
matrix = [list(map(int, input())) for _ in range(N)]

print(bfs())

File: backup/Breadth-first_search(BFS)/c++/05_미로탈출.cpp
// 이것이 취업을 위한 코딩테스트다 152p
// https://www.acmicpc.net/problem/2178
// 2

#include <iostream>
#include <queue>
using namespace std;

int m, n;
int matrix[201][201];

int mx[4] = { -1, 1, 0, 0 };
int my[4] = { 0, 0, -1, 1 };

int bfs(int x, int y)
{
	queue<pair<int, int>> q;
	q.push({ x, y });

	while (!q.empty())
	{
		int dx = q.front().first;
		int dy = q.front().second;
		q.pop();

		for (int i = 0; i < 4; i++)
		{
			int nx = dx + mx[i];
			int ny = dy + my[i];
			if (nx < 0 || nx >= n || ny < 0 || ny >= m)
				continue;
			if (matrix[nx][ny] == 0)
				continue;
			if (matrix[nx][ny] == 1)
			{
				matrix[nx][ny] = matrix[dx][dy] + 1;
				q.push({ nx, ny });
			}
		}
	}
	return matrix[n - 1][m - 1];
}

int main()
{
	cin >> n >> m;

	for (int i = 0; i < n; i++)
	{
		for (int j = 0; j < m; j++)
			scanf("%1d", &matrix[i][j]);
	}

	int result = bfs(0, 0);
	cout << result << '\n';
	return 0;
}